File: java/data-structure/src/main/java/org/example/HashTrieTree.java
package org.example;

/**
 * 哈希实现字典树
 */
public class HashTrieTree implements TrieTree {
    private final Node root = new Node('\0');
    private int size;

    @Override
    public int size() {
        return size;
    }

    @Override
    public boolean isEmpty() {
        return size() == 0;
    }

    @Override
    public boolean add(String word) {
        if (word == null) {
            throw new RuntimeException("必须有");
        }
        Node curr = root;
        for (char ch : word.toCharArray()) {
            Node next = curr.map.get(ch);
            if (next == null) {
                next = new Node(ch);
                curr.map.put(ch, next);
            }
            curr = next;
        }
        if (curr.isWord) {
            return false;
        }
        curr.isWord = true;
        size++;
        return true;
    }

    @Override
    public boolean contains(String word) {
        if (word == null) {
            throw new RuntimeException("必须有");
        }
        Node node = getNode(word);
        return node != null && node.isWord;
    }

    @Override
    public boolean remove(String word) {
        if (word == null) {
            throw new RuntimeException("必须有");
        }
        Node curr = getNode(word);
        if (curr == null || !curr.isWord) {
            return false;
        }
        curr.isWord = false;
        size--;
        clear(root, false);
        return true;
    }

    @Override
    public void clear() {
        clear(root, true);
        root.isWord = false;
        size = 0;
    }

    @Override
    public boolean containsPrefix(String prefix) {
        if (prefix == null) {
            throw new RuntimeException("必须有");
        }
        return getNode(prefix) != null;
    }

    @Override
    public boolean removePrefix(String prefix) {
        if (prefix == null) {
            throw new RuntimeException("必须有");
        }
        Node node = getNode(prefix);
        if (node == null) {
            return false;
        }
        clear(node, true);
        if (node.isWord) {
            node.isWord = false;
            size--;
        }
        clear(root, false);
        if (root.isWord && "".equals(prefix)) {
            root.isWord = false;
            size--;
        }
        return true;
    }

    @Override
    public Set<String> matchPrefix(String prefix) {
        if (prefix == null) {
            throw new RuntimeException("必须有");
        }
        return buildWord(getNode(prefix), prefix);
    }

    @Override
    public Iterator<String> iterator() {
        return buildWord(root, "").iterator();
    }

    private Node getNode(String prefix) {
        Node curr = root;
        for (char c : prefix.toCharArray()) {
            Node next = curr.map.get(c);
            if (next == null) {
                return null;
            }
            curr = next;
        }
        return curr;
    }

    private boolean clear(Node curr, boolean isDelete) {
        Iterator<Node> it = curr.map.values().iterator();
        if (isDelete) {
            while (it.hasNext()) {
                Node next = it.next();
                if (next.isWord) {
                    size--;
                    next.isWord = false;
                }
                clear(next, true);
            }
        } else {
            while (it.hasNext()) {
                if (clear(it.next(), false)) {
                    return true;
                }
            }
            if (curr.isWord) {
                return true;
            }
        }
        curr.map.clear();
        return false;
    }

    private Set<String> buildWord(Node curr, String prefix) {
        Set<String> set = new HashSet<>();
        if (curr == null) {
            return set;
        }
        if (curr.isWord) {
            set.add(prefix);
        }
        StringBuilder sb = new StringBuilder();
        Iterator<Node> it = curr.map.values().iterator();
        while (it.hasNext()) {
            buildWord(it.next(), prefix, sb, set);
        }
        return set;
    }

    private void buildWord(Node curr, String prefix, StringBuilder sb, Set<String> set) {
        sb.append(curr.ch);
        if (curr.isWord) {
            set.add(prefix + sb);
        }
        Iterator<Node> it = curr.map.values().iterator();
        while (it.hasNext()) {
            buildWord(it.next(), prefix, sb, set);
        }
        sb.deleteCharAt(sb.length() - 1);
    }

    @Override
    public Set<String> intersection(Set<? extends String> other) {
        if (other == null) {
            throw new RuntimeException("必须有");
        }
        Set<String> set = new HashSet<>();
        Iterator<? extends String> it = other.iterator();
        while (it.hasNext()) {
            String next = it.next();
            if (contains(next)) {
                set.add(next);
            }
        }
        return set;
    }

    @Override
    public Set<String> union(Set<? extends String> other) {
        if (other == null) {
            throw new RuntimeException("必须有");
        }
        Set<String> set = new HashSet<>();
        Iterator<String> it1 = iterator();
        while (it1.hasNext()) {
            set.add(it1.next());
        }
        Iterator<? extends String> it2 = other.iterator();
        while (it2.hasNext()) {
            set.add(it2.next());
        }
        return set;
    }

    @Override
    public Set<String> complement(Set<? extends String> other) {
        if (other == null) {
            throw new RuntimeException("必须有");
        }
        Iterator<? extends String> it2 = other.iterator();
        while (it2.hasNext()) {
            if (!contains(it2.next())) {
                return new HashSet<>();
            }
        }
        return difference(other);
    }

    @Override
    public Set<String> difference(Set<? extends String> other) {
        if (other == null) {
            throw new RuntimeException("必须有");
        }
        Set<String> set = new HashSet<>();
        Iterator<String> it1 = iterator();
        while (it1.hasNext()) {
            set.add(it1.next());
        }
        Iterator<? extends String> it2 = other.iterator();
        while (it2.hasNext()) {
            set.remove(it2.next());
        }
        return set;
    }

    @Override
    public Set<String> symmetric(Set<? extends String> other) {
        if (other == null) {
            throw new RuntimeException("必须有");
        }
        Set<String> set = union(other);
        Iterator<String> it = intersection(other).iterator();
        while (it.hasNext()) {
            set.remove(it.next());
        }
        return set;
    }

    private static class Node {
        private final Map<Character, Node> map = new HashMap<>();
        private final char ch;
        private boolean isWord;

        private Node(char ch) {
            this.ch = ch;
        }
    }
}

File: csharp/data-structure/HashTrieTree.cs
using System.Text;

namespace Org.Example
{
    public class HashTrieTree : ITrieTree
    {
        private readonly Node root = new Node('\0');

        public int Count { get; private set; }

        public bool IsEmpty => Count == 0;

        public void Clear()
        {
            Clear(root, true);
            root.isWord = false;
            Count = 0;
        }

        public bool Add(string item)
        {
            if (item == default)
            {
                throw new Exception("必须有");
            }
            Node curr = root;
            foreach (char ch in item)
            {
                Node next = curr.map[ch];
                if (next == default)
                {
                    next = new Node(ch);
                    curr.map[ch] = next;
                }
                curr = next;
            }
            if (curr.isWord)
            {
                return false;
            }
            curr.isWord = true;
            Count++;
            return true;
        }

        public bool Contains(string item)
        {
            if (item == default)
            {
                throw new Exception("必须有");
            }
            Node node = GetNode(item);
            return node != default && node.isWord;
        }

        public bool Remove(string item)
        {
            if (item == default)
            {
                throw new Exception("必须有");
            }
            Node curr = GetNode(item);
            if (curr == default || !curr.isWord)
            {
                return false;
            }
            curr.isWord = false;
            Count--;
            Clear(root, false);
            return true;
        }

        public bool Intersection(ISet<string> other)
        {
            if (other == default)
            {
                throw new Exception("必须有");
            }
            bool flag = false;
            foreach (var item in this)
            {
                if (!other.Contains(item))
                {
                    flag |= Remove(item);
                }
            }
            return flag;
        }

        public bool Union(ISet<string> other)
        {
            if (other == default)
            {
                throw new Exception("必须有");
            }
            bool flag = false;
            foreach (var item in other)
            {
                flag |= Add(item);
            }
            return flag;
        }

        public bool Complement(ISet<string> other)
        {
            if (other == default)
            {
                throw new Exception("必须有");
            }
            foreach (var item in other)
            {
                if (!Contains(item))
                {
                    return false;
                }
            }
            return Difference(other);
        }

        public bool Difference(ISet<string> other)
        {
            if (other == default)
            {
                throw new Exception("必须有");
            }
            bool flag = false;
            foreach (var item in other)
            {
                flag |= Remove(item);
            }
            return flag;
        }

        public bool Symmetric(ISet<string> other)
        {
            if (other == default)
            {
                throw new Exception("必须有");
            }
            bool flag = false;
            foreach (var item in other)
            {
                flag |= Contains(item) ? Remove(item) : Add(item);
            }
            return flag;
        }

        public bool ContainsPrefix(string prefix)
        {
            if (prefix == default)
            {
                throw new Exception("必须有");
            }
            return GetNode(prefix) != default;
        }

        public bool RemovePrefix(string prefix)
        {
            if (prefix == default)
            {
                throw new Exception("必须有");
            }
            Node node = GetNode(prefix);
            if (node == default)
            {
                return false;
            }
            Clear(node, true);
            if (node.isWord)
            {
                node.isWord = false;
                Count--;
            }
            Clear(root, false);
            if (root.isWord && "" == prefix)
            {
                root.isWord = false;
                Count--;
            }
            return true;
        }

        public ISet<string> MatchPrefix(string prefix)
        {
            if (prefix == default)
            {
                throw new Exception("必须有");
            }
            return BuildWord(GetNode(prefix), prefix);
        }

        public IEnumerator<string> GetEnumerator() => MatchPrefix("").GetEnumerator();

        private Node GetNode(string prefix)
        {
            Node curr = root;
            foreach (char c in prefix)
            {
                Node next = curr.map[c];
                if (next == default)
                {
                    return default;
                }
                curr = next;
            }
            return curr;
        }

        private bool Clear(Node curr, bool isDelete)
        {
            if (isDelete)
            {
                foreach (var item in curr.map.Values)
                {
                    if (item.isWord)
                    {
                        Count--;
                        item.isWord = false;
                    }
                    Clear(item, true);
                }
            }
            else
            {
                foreach (var item in curr.map.Values)
                {
                    if (Clear(item, false))
                    {
                        return true;
                    }
                }
                if (curr.isWord)
                {
                    return true;
                }
            }
            curr.map.Clear();
            return false;
        }

        private ISet<string> BuildWord(Node curr, string prefix)
        {
            ISet<string> set = new HashSet<string>();
            if (curr == default)
            {
                return set;
            }
            if (curr.isWord)
            {
                set.Add(prefix);
            }
            StringBuilder sb = new StringBuilder();
            foreach (var item in curr.map.Values)
            {
                BuildWord(item, prefix, sb, set);
            }
            return set;
        }

        private void BuildWord(Node curr, string prefix, StringBuilder sb, ISet<string> set)
        {
            sb.Append(curr.ch);
            if (curr.isWord)
            {
                set.Add(prefix + sb);
            }
            foreach (var item in curr.map.Values)
            {
                BuildWord(item, prefix, sb, set);
            }
            sb.Remove(sb.Length - 1, 1);
        }

        private class Node
        {
            public readonly IMap<char, Node> map = new HashMap<char, Node>();
            public readonly char ch;
            public bool isWord;

            public Node(char ch) => this.ch = ch;
        }
    }
}